src/ophyd_async/panda/writers/hdf_writer.py

```diff
@@ -0,0 +1,215 @@
+import asyncio
+from enum import Enum
+from pathlib import Path
+from typing import Any, AsyncGenerator, AsyncIterator, Dict, List, Optional, Union
+
+from bluesky.protocols import Descriptor, StreamAsset
+from p4p.client.thread import Context
+
+from ophyd_async.core import (
+    DEFAULT_TIMEOUT,
+    DetectorWriter,
+    Device,
+    DirectoryProvider,
+    NameProvider,
+    SignalR,
+    wait_for_value,
+)
+from ophyd_async.core.signal import observe_value
+from ophyd_async.panda.panda import PandA
+
+from .panda_hdf_file import _HDFDataset, _HDFFile
+
+
+class Capture(str, Enum):
+    # Capture signals for the HDF Panda
+    No = "No"
+    Value = "Value"
+    Diff = "Diff"
+    Sum = "Sum"
+    Mean = "Mean"
+    Min = "Min"
+    Max = "Max"
+    MinMax = "Min Max"
+    MinMaxMean = "Min Max Mean"
+
+
+def get_capture_signals(
+    block: Device, path_prefix: Optional[str] = ""
+) -> Dict[str, SignalR]:
+    """Get dict mapping a capture signal's name to the signal itself"""
+    if not path_prefix:
+        path_prefix = ""
+    signals: Dict[str, SignalR[Any]] = {}
+    for attr_name, attr in block.children():
+        # Capture signals end in _capture, but num_capture is a red herring
+        if attr_name == "num_capture":
+            continue
+        dot_path = f"{path_prefix}{attr_name}"
+        if isinstance(attr, SignalR) and attr_name.endswith("_capture"):
+            signals[dot_path] = attr
+        attr_signals = get_capture_signals(attr, path_prefix=dot_path + ".")
+        signals.update(attr_signals)
+    return signals
+
+
+# This should return a dictionary which contains a dict, containing the Capture
+# signal object, and the value of that signal
+async def get_signals_marked_for_capture(
+    capture_signals: Dict[str, SignalR]
+) -> Dict[str, Dict[str, Union[SignalR, Capture]]]:
+
+    # Read signals to see if they should be captured
+    do_read = [signal.get_value() for signal in capture_signals.values()]
+
+    signal_values = await asyncio.gather(*do_read)
+
+    assert len(signal_values) == len(
+        capture_signals
+    ), "Length of read signals are different to length of signals"
+
+    signals_to_capture: Dict[str, Dict[str, Union[SignalR, Capture]]] = {}
+    for signal_path, signal_object, signal_value in zip(
+        capture_signals.keys(), capture_signals.values(), signal_values
+    ):
+
+        signal_path = signal_path.replace("_capture", "")
+        if (signal_value.value in iter(Capture)) and (signal_value.value != Capture.No):
+            signals_to_capture[signal_path] = {
+                "signal": signal_object,
+                "capture_type": signal_value.value,
+            }
```

Comment:
Please can you put this in either a dataclass or a plain tuple? Otherwise we lose type information here

Reply:
Using this now:

https://github.com/bluesky/ophyd-async/blob/7e328e7462dd178b3d0451da3947f803421bcb9e/src/ophyd_async/panda/writers/hdf_writer.py#L57-L60